agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,1082 @@
+		c.state = newState
+	} else {
+		return fmt.Errorf("CA is already in state %q", state)
+	}
+	return nil
+}
+
+// setPrimaryRoots updates the most recently seen roots from the primary.
+func (c *CAManager) setPrimaryRoots(newRoots structs.IndexedCARoots) error {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+
+	if c.state == CAStateInitializing || c.state == CAStateReconfig {
+		c.primaryRoots = newRoots
+	} else {
+		return fmt.Errorf("Cannot update primary roots in state %q", c.state)
+	}
+	return nil
+}
+
+func (c *CAManager) getPrimaryRoots() structs.IndexedCARoots {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+	return c.primaryRoots
+}
+
+// initializeCAConfig is used to initialize the CA config if necessary
+// when setting up the CA during establishLeadership
+func (c *CAManager) initializeCAConfig() (*structs.CAConfiguration, error) {
+	state := c.srv.fsm.State()
+	_, config, err := state.CAConfig(nil)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		config = c.srv.config.CAConfig
+		if config.ClusterID == "" {
+			id, err := uuid.GenerateUUID()
+			if err != nil {
+				return nil, err
+			}
+			config.ClusterID = id
+		}
+	} else if _, ok := config.Config["IntermediateCertTTL"]; !ok {
+		dup := *config
+		copied := make(map[string]interface{})
+		for k, v := range dup.Config {
+			copied[k] = v
+		}
+		copied["IntermediateCertTTL"] = connect.DefaultIntermediateCertTTL.String()
+		dup.Config = copied
+		config = &dup
+	} else {
+		return config, nil
+	}
+
+	req := structs.CARequest{
+		Op:     structs.CAOpSetConfig,
+		Config: config,
+	}
+	if resp, err := c.srv.raftApply(structs.ConnectCARequestType, req); err != nil {
+		return nil, err
+	} else if respErr, ok := resp.(error); ok {
+		return nil, respErr
+	}
+
+	return config, nil
+}
+
+// parseCARoot returns a filled-in structs.CARoot from a raw PEM value.
+func parseCARoot(pemValue, provider, clusterID string) (*structs.CARoot, error) {
+	id, err := connect.CalculateCertFingerprint(pemValue)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing root fingerprint: %v", err)
+	}
+	rootCert, err := connect.ParseCert(pemValue)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing root cert: %v", err)
+	}
+	keyType, keyBits, err := connect.KeyInfoFromCert(rootCert)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting root key info: %v", err)
+	}
+	return &structs.CARoot{
+		ID:                  id,
+		Name:                fmt.Sprintf("%s CA Root Cert", strings.Title(provider)),
+		SerialNumber:        rootCert.SerialNumber.Uint64(),
+		SigningKeyID:        connect.EncodeSigningKeyID(rootCert.SubjectKeyId),
+		ExternalTrustDomain: clusterID,
+		NotBefore:           rootCert.NotBefore,
+		NotAfter:            rootCert.NotAfter,
+		RootCert:            pemValue,
+		PrivateKeyType:      keyType,
+		PrivateKeyBits:      keyBits,
+		Active:              true,
+	}, nil
+}
+
+// getCAProvider is being called while holding caProviderReconfigurationLock
```

Comment:
I think this comment is stale now, since it references `caProviderReconfigurationLock`